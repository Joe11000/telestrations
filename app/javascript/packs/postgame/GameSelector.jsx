import React from 'react';
import PropTypes from 'prop-types';

export default class GameSelector extends React.Component {
  constructor(props){
    super(props);
  }
  handleChangeGameSelector = event => {
    event.preventDefault();
    const game_id = parseInt(event.target.value);
    this.props.retrieveCardsForPostgame(game_id);
  }

  render() {
    const {all_postgames_of__current_user, current_postgame_id} = this.props;
    
    
    // debugger
    return (
      <React.Fragment>
        {
          all_postgames_of__current_user &&
          current_postgame_id &&
          <select value={current_postgame_id} className='custom-select' onChange={this.handleChangeGameSelector}>
            {
              all_postgames_of__current_user.map(function(game_info, index) {
                return(<option key={game_info.id} value={game_info.id}>Game {index + 1} - {game_info.created_at_strftime}</option>);
              })
            }
          </select>
        }
      </React.Fragment>
    )
  }
}



GameSelector.propTypes = {
  all_postgames_of__current_user: PropTypes.arrayOf(function(propValue, key, componentName, location, propFullName) {
    let _propValueTypesValidator = {
      'id': 'number',
      'created_at_strftime': 'string'
    }

    Object.keys(_propValueTypesValidator).forEach(propValueKey => {
      if(typeof propValue[key][propValueKey] != _propValueTypesValidator[propValueKey]) {
        return new Error(
          'Invalid prop `' + propFullName + '` supplied to' +
          ' `' + componentName + '`. Validation failed.'
        )
      }
    })
  }).isRequired,

  current_postgame_id: PropTypes.number.isRequired,
  retrieveCardsForPostgame: PropTypes.func.isRequired,
}
