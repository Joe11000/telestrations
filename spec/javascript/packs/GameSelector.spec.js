import React from 'react'
import {shallow} from 'enzyme'
import GameSelector from 'packs/postgame/GameSelector'
import {render, fireEvent} from 'react-testing-library'

import enzymeSerializer from 'enzyme-to-json/serializer'
expect.addSnapshotSerializer(enzymeSerializer)

describe('GameSelector Component', ()=>{
  it('has propTypes', ()=>{

    // props = {
    //           all_postgames_of__current_user: PropTypes.shape({
    //             id: PropTypes.number.isRequired,
    //             created_at_strftime: PropTypes.string.isRequired
    //           }),
    //           retrieveCardsForPostgame: PropTypes.func.isRequired

    //         }
  })

  describe('sets correct values', ()=>{
    it('onload', ()=>{
      let game_1 = {'id': 11, 'created_at_strftime': 'Mon Nov 1, 2018'}
      let game_2 = {'id': 22, 'created_at_strftime': 'Tues Nov 2, 2018'}
      let game_3 = {'id': 33, 'created_at_strftime': 'Wed Nov 3, 2018'}

      let props = {
                      'all_postgames_of__current_user': [
                                                          {
                                                            'id': game_1.id,
                                                            'created_at_strftime': game_1.created_at_strftime
                                                          },
                                                          {
                                                            'id': game_2.id,
                                                            'created_at_strftime': game_2.created_at_strftime
                                                          },
                                                          {
                                                            'id': game_3.id,
                                                            'created_at_strftime': game_3.created_at_strftime
                                                          }
                                                        ],

                      'retrieveCardsForPostgame': jest.fn()
                    }

      const game_selector = shallow(<GameSelector {...props} />)

      expect( game_selector).toMatchSnapshot();
    });

    test.only('selector onChange calls props.retrieveCardsForPostgame(game_id)', ()=>{
      let game_1 = {'id': 11, 'created_at_strftime': 'Mon Nov 1, 2018'}
      let game_2 = {'id': 22, 'created_at_strftime': 'Tues Nov 2, 2018'}
      let game_3 = {'id': 33, 'created_at_strftime': 'Wed Nov 3, 2018'}

      const mockRetrieveCardsForPostgame = jest.fn()

      let props = {
                      'all_postgames_of__current_user': [
                                                          {
                                                            'id': game_1.id,
                                                            'created_at_strftime': game_1.created_at_strftime
                                                          },
                                                          {
                                                            'id': game_2.id,
                                                            'created_at_strftime': game_2.created_at_strftime
                                                          },
                                                          {
                                                            'id': game_3.id,
                                                            'created_at_strftime': game_3.created_at_strftime
                                                          }
                                                        ],

                      'retrieveCardsForPostgame': mockRetrieveCardsForPostgame
                    }

      const game_selector = shallow(<GameSelector {...props} />)

      const change_to_value = `${game_1.id}`
      let params = {'target': {'value': change_to_value }, "preventDefault": ()=>{} }
      game_selector.find('select').simulate('change',  params )

      expect(mockRetrieveCardsForPostgame.mock.calls.length).toBe(1)
      expect(mockRetrieveCardsForPostgame).toBeCalledWith(game_1.id)
    });
  });
});
