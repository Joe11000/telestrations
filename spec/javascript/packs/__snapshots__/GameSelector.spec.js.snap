// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameSelector Component sets correct values 1`] = `
<div
  id="game-selector-container"
>
  <select
    className="custom-select"
    onChange={[Function]}
  >
    <option
      key="11"
      value={11}
    >
      Game 
      1
       - 
      Mon Nov 1, 2018
    </option>
    <option
      key="22"
      value={22}
    >
      Game 
      2
       - 
      Tues Nov 2, 2018
    </option>
    <option
      key="33"
      value={33}
    >
      Game 
      3
       - 
      Wed Nov 3, 2018
    </option>
  </select>
</div>
`;

exports[`GameSelector Component sets correct values onload 1`] = `
<div
  id="game-selector-container"
>
  <select
    className="custom-select"
    onChange={[Function]}
  >
    <option
      key="11"
      value={11}
    >
      Game 
      1
       - 
      Mon Nov 1, 2018
    </option>
    <option
      key="22"
      value={22}
    >
      Game 
      2
       - 
      Tues Nov 2, 2018
    </option>
    <option
      key="33"
      value={33}
    >
      Game 
      3
       - 
      Wed Nov 3, 2018
    </option>
  </select>
</div>
`;
